pkg/generate/app/cmd/newapp.go

```diff
@@ -279,36 +292,68 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
 		common := app.PipelineGroup{}
 		for _, ref := range group {
 
-			var pipeline *app.Pipeline
-			if ref.Input().ExpectToBuild {
-				glog.V(2).Infof("will use %q as the base image for a source build of %q", ref, ref.Input().Uses)
-				input, err := app.InputImageFromMatch(ref.Input().Match)
-				if err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
-				}
-				strategy, source, err := app.StrategyAndSourceForRepository(ref.Input().Uses)
-				if err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
-				}
-				if pipeline, err = app.NewBuildPipeline(ref.Input().String(), input, strategy, source); err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
+			// TODO template and other individual matches are mutually exclusive,
+			// need to investigate the side-effects of having them together
+			if ref.Input().Match.Template != nil {
+				glog.V(2).Infof("will include objects from template %q", ref)
```

Comment:
Can you describe a scenario where it would make sense for a reasonably complex template to be paired with the input of other components?

I.e.:

    # mysql-cluster is a template
    $ osc new-app php~somegitrepo mysql-cluster
    # no overlap between components

    # mysql-cluster is a template, mysql is an image
    $ osc new-app mysql-cluster mysql
    # doesn't seem like the two should overlap, but they might conflict on their service name

The only case I can think of is when you have a template and a git repo:

    # eap is a template with a build config
    $ osc new-app <gitrepo> eap

In this case, the "eap" template would have to define a build config (or at least an input image) and we'd need to know to link those up.  So EAP's pipeline *might* be mutated to take the git repo as input... but that seems like a special case.

If the user specifies a single template, I don't think we should be breaking out the pipelines.  In the case above, it seems almost like a special case.  I would say that it's possible that templates should be given a chance to absorb arguments prior to execution, but then are added after the pipelines are joined.  Templates should probably be kept together as long as possible, but we should identify conflicts with the other arguments as soon as possible.

Honestly, at this point it might be better to disallow any mixing, and simply focus on getting the validation and checking behavior right.

----- Original Message -----
> > -			if ref.Input().ExpectToBuild {
> > -				glog.V(2).Infof("will use %q as the base image for a source build of
> > %q", ref, ref.Input().Uses)
> > -				input, err := app.InputImageFromMatch(ref.Input().Match)
> > -				if err != nil {
> > -					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
> > -				}
> > -				strategy, source, err :=
> > app.StrategyAndSourceForRepository(ref.Input().Uses)
> > -				if err != nil {
> > -					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
> > -				}
> > -				if pipeline, err = app.NewBuildPipeline(ref.Input().String(), input,
> > strategy, source); err != nil {
> > -					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
> > +			// TODO template and other individual matches are mutually exclusive,
> > +			// need to investigate the side-effects of having them together
> > +			if ref.Input().Match.Template != nil {
> > +				glog.V(2).Infof("will include objects from template %q", ref)
> 
> @smarterclayton I wasn't sure about the correct strategy here:
> 
> 1. Break down the template and handle each of its objects as one individual
> pipeline here in `buildPipelines`. I think this is the right approach, but I
> will need to figure out how to handle Services, Routes, etc.
> 2. Break down the template before, maybe in `validate` and make each separate
> object a `ComponentInput` (I think it's wrong).
> 3. Handle the Template as one single pipeline element returned by
> `buildPipelines` and break it down later.
> 
> Thoughts?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1716/files#r28283935